### ISSUE TITLE
`dependencySerializationTraitProperty.unsupportedReadOnlyProperty` is a false positive — readonly properties survive `serialize()`/`unserialize()` fine

### ISSUE THREAD

**Issue by Niklan** [mglaman/phpstan-drupal#1009]
# Bug report

- Version: 2.1.0

`DependencySerializationTraitPropertyRule` unconditionally flags any `readonly` property on a class using `Drupal\Core\DependencyInjection\DependencySerializationTrait`, regardless of the property's type or whether the class participates in an inheritance chain:

https://github.com/mglaman/phpstan-drupal/blob/cab71f31fae33981446137738e1709fc8c3ea2b8/src/Rules/Drupal/DependencySerializationTraitPropertyRule.php#L42-L51

On current Drupal core (verified on 11.3.14 / PHP 8.4), this is a false positive for the common case of a class declaring its own `readonly` constructor-promoted service property.

`DependencySerializationTrait::__sleep()` does not mutate the property itself. It only removes the property name from the array of names returned to `serialize()`:

```php
elseif ($service_id = $reverse_container->getId($value)) {
    $this->_serviceIds[$key] = $service_id;
    unset($vars[$key]);
}
```

`__wakeup()` then assigns the restored service to the property on the newly unserialized object, where the property was never previously initialized (it was excluded from serialization). This is the first write to the property on that instance, from within the class's own declaring scope — exactly what PHP's `readonly` semantics allow.

### Code snippet that reproduces the problem

Script for `drush scr`:

```php
<?php

use Drupal\Core\Database\Connection;
use Drupal\Core\DependencyInjection\DependencySerializationTrait;

final class ReadonlyServiceHolder {

  use DependencySerializationTrait;

  public function __construct(
    protected readonly Connection $database,
    protected readonly string $foo,
  ) {}

  public function ping(): string {
    return (string) $this->database->query('SELECT 1')->fetchField();
  }

}

$holder = new ReadonlyServiceHolder(\Drupal::service('database'), 'bar');

echo 'Before serialize(): ' . $holder->ping() . PHP_EOL;

$serialized = serialize($holder);
echo 'Serialized payload (no Connection/PDO object inside, only a service id):' . PHP_EOL;
echo $serialized . PHP_EOL . PHP_EOL;

/** @var ReadonlyServiceHolder $restored */
$restored = unserialize($serialized);

echo 'After unserialize() + __wakeup(): ' . $restored->ping() . PHP_EOL;
echo 'SUCCESS: readonly Connection property survived the sleep/wakeup cycle.' . PHP_EOL;
```

Actual output:

```
Before serialize(): 1
Serialized payload (no Connection/PDO object inside, only a service id):
O:21:"ReadonlyServiceHolder":2:{s:14:"*_serviceIds";a:1:{s:8:"database";s:8:"database";}s:18:"*_entityStorages";a:0:{}}

After unserialize() + __wakeup(): 1
SUCCESS: readonly Connection property survived the sleep/wakeup cycle.
```

PHPStan's output:

```
  11     Read-only properties are incompatible with Drupal\Core\DependencyInjection\DependencySerializationTrait.  
         🪪  dependencySerializationTraitProperty.unsupportedReadOnlyProperty                                      
         💡  See https://www.drupal.org/node/3110266                                                               
  12     Read-only properties are incompatible with Drupal\Core\DependencyInjection\DependencySerializationTrait.  
         🪪  dependencySerializationTraitProperty.unsupportedReadOnlyProperty                                      
         💡  See https://www.drupal.org/node/3110266    
```

Note how it is also triggered for `$foo` property, which is not affected by the trait at all.

At minimum, the `readonly` check shouldn't fire unconditionally:

- Scalar/array/non-object `readonly` properties are never touched by `__sleep()` at all (`if (!is_object($value)) continue;`), so flagging them is unconditionally wrong.
- For object-typed properties resolvable via the container, the sleep/wakeup swap logic (shown above) works correctly with `readonly` in the current core implementation — the rule doesn't appear to reflect this and may be based on an older `__sleep()` implementation that mutated the live property directly.

The only requirement for this trait is that the property holding a service must be `public` or `protected`. The `readonly` requirement is not needed.

So it looks to me like the whole `dependencySerializationTraitProperty.unsupportedReadOnlyProperty` rule is just broken and should be removed? Maybe I'm missing some of its useful checks, but in its current state it simply creates noise for no reason and forces changes to code that isn't broken at all.

Probably related: https://github.com/mglaman/phpstan-drupal/issues/889

**Comment by mglaman** [comment]
Confirmed. `__sleep()` never mutates the property — it only drops the name from the array handed to `serialize()` — and `__wakeup()` performs the first write on the freshly unserialized object from the composing class scope, which readonly allows. Flagging every readonly property was wrong, and scalar/array properties were never touched by the trait in the first place.

The rule isn't removed entirely, though. There is one real failure mode: the trait used by a parent class with the readonly property declared in the child. `__wakeup()` then runs in the parent scope and cannot initialize the child's readonly property. Reduced test case, no Drupal required:

```php
trait T {
    public function __sleep(): array
    {
        return array_diff(array_keys(get_object_vars($this)), ['db']);
    }

    public function __wakeup(): void
    {
        $this->db = 'restored';
    }
}

class DirectUse {
    use T;
    public function __construct(public readonly string $db) {}
}

class ParentUses {
    use T;
}
class Child extends ParentUses {
    public function __construct(public readonly string $db) {}
}
```

```
PHP 8.3.32
DirectUse: OK, db=restored
Child: ERROR: Cannot initialize readonly property Child::$db from scope ParentUses
PHP 8.4.23
DirectUse: OK, db=restored
Child: OK, db=restored
```

That child case is exactly the common Drupal pattern of a form extending `FormBase` (which uses the trait) with readonly promoted service properties. PHP 8.4 relaxed the initialization scope, so it only breaks on 8.3 and earlier.

#1010 narrows the rule to report only that case: object-typed readonly property, trait inherited from a parent class, `phpVersion` below 8.4. Your `ReadonlyServiceHolder` example passes clean, including the `$foo` string property. The error identifier is unchanged so existing baselines and ignores keep working.
